include/flatbuffers/grpc.h

```diff
@@ -22,49 +22,198 @@
 #include "grpc++/support/byte_buffer.h"
 #include "grpc/byte_buffer_reader.h"
 
+namespace flatbuffers {
+namespace grpc {
+
+enum AddRef { ADD_REF };
+enum StealRef { STEAL_REF };
+
+template <class T>
+class Message {
+ public:
+  inline Message()
+      : slice_(grpc_empty_slice()) {
+  }
+
+  inline Message(grpc_slice slice, AddRef)
+      : slice_(grpc_slice_ref(slice)) {
+  }
+
+  inline Message(grpc_slice slice, StealRef)
+      : slice_(slice) {
+  }
+
+  inline Message(const Message &other)
+      : slice_(grpc_slice_ref(other.slice_)) {
+  }
+
+  inline Message(Message &&other)
+      : slice_(other.slice_) {
+    other.slice_ = grpc_empty_slice();
+  }
+
+  inline Message &operator=(const Message &other) {
+    slice_ = grpc_slice_ref(other.slice_);
+  }
+
+  inline Message &operator=(Message &&other) {
+    slice_ = other.slice_;
+    other.slice_ = grpc_empty_slice();
+  }
+
+  inline ~Message() {
+    grpc_slice_unref(slice_);
+  }
+
+  inline const grpc_slice& BorrowSlice() const {
+    return slice_;
+  }
+
+  inline bool Verify() const {
+    if (slice_.refcount != nullptr) {
+      Verifier verifier(slice_.data.refcounted.bytes, slice_.data.refcounted.length);
+      return verifier.VerifyBuffer<T>(nullptr);
+    } else {
+      Verifier verifier(slice_.data.inlined.bytes, slice_.data.inlined.length);
+      return verifier.VerifyBuffer<T>(nullptr);
+    }
+  }
+
+  inline const T* GetRoot() const {
+    if (slice_.refcount != nullptr) {
+      return flatbuffers::GetRoot<T>(slice_.data.refcounted.bytes);
+    } else {
+      return flatbuffers::GetRoot<T>(slice_.data.inlined.bytes);
+    }
+  }
+
+ protected:
+  grpc_slice slice_;
+};
+
+// This isnt a true allocator, as it only allows up to two oustanding
+// allocation at a time. However, this is sufficient for our use case, and
+// simplifies the grpc_slice tracking.
+class SliceAllocator : public simple_allocator {
+ public:
+  inline SliceAllocator()
+      : slices_{grpc_empty_slice(), grpc_empty_slice()} {
+  }
+
+  SliceAllocator(const SliceAllocator& other) = delete;
+  SliceAllocator& operator=(const SliceAllocator& other) = delete;
+
+  virtual inline ~SliceAllocator() {
+    grpc_slice_unref(slices_[0]);
+    grpc_slice_unref(slices_[1]);
+  }
+
+  virtual inline uint8_t *allocate(size_t size) const override {
+    if (GRPC_SLICE_IS_EMPTY(slices_[0])) {
```

Comment:
It's only intended to be used by flatbuffer's internal `vector_downward`, and there there is one-to-one correspondence between them (i.e. a single vector owner per allocator) due to the copy ctors being disabled. It needs two because `reallocate` has one in flight while it does the `memcpy`. 

In theory the allocator should only need one `grpc_slice`, and if we are up for changing the implementation a bit, it would simplify things to shift that into a `reallocate` func on the allocator that takes care of the memcpy and such. Let me get some of the other PR's sorted out and I'll update this one.

If we are worried about people seeing/using the allocator without knowing it's specific limitations, we could plop it into a `flatbuffers::grpc::details` namespace or something, or disable the ctor outside of predefined friend classes.